File: src/main/services/ModArchiveExtractor.ts
import fs from 'fs'
import path from 'path'
import { spawn } from 'child_process'
import { app } from 'electron'
import AdmZip from 'adm-zip'
import { path7za } from '7zip-bin'
import { GMP_METADATA_FOLDER, GMP_FALLBACK_METADATA_FOLDER, TEMP_FOLDER } from '../constants'
import { isInside, safeZipEntryPath } from './PathGuard'
import { debugLog } from '../logger'


function isZipFormat(archivePath: string): boolean {
  const lower = archivePath.toLowerCase()
  if (
    lower.endsWith('.zip') ||
    lower.endsWith('.jar') ||
    lower.endsWith('.apk') ||
    lower.endsWith('.bbmod') ||
    lower.endsWith('.gmp')
  ) {
    return true
  }
  return startsWithZipMagic(archivePath)
}

function startsWithZipMagic(archivePath: string): boolean {
  try {
    const fd = fs.openSync(archivePath, 'r')
    const buf = Buffer.alloc(4)
    let n = 0
    try {
      n = fs.readSync(fd, buf, 0, 4, 0)
    } finally {
      fs.closeSync(fd)
    }
    return n >= 2 && buf[0] === 0x50 && buf[1] === 0x4b
  } catch {
    return false
  }
}









function resolve7za(): string {
  const arch = process.arch === 'ia32' ? 'ia32' : process.arch === 'arm64' ? 'arm64' : 'x64'
  const candidates: string[] = []

  if (process.platform === 'win32') {
    const bundled7z = app.isPackaged
      ? path.join(process.resourcesPath, '7z', '7z.exe')
      : path.join(app.getAppPath(), 'resources', '7z', '7z.exe')
    if (fs.existsSync(bundled7z)) candidates.push(bundled7z)
    const localAppData = process.env['LOCALAPPDATA'] || ''
    const programFiles = process.env['ProgramFiles'] || ''
    const programFilesX86 = process.env['ProgramFiles(x86)'] || ''
    candidates.push(
      path.join(localAppData, 'Microsoft', 'WindowsApps', '7z.exe'),
      path.join(programFiles, '7-Zip', '7z.exe'),
      path.join(programFilesX86, '7-Zip', '7z.exe')
    )
  }

  if (app.isPackaged) candidates.unshift(path.join(process.resourcesPath, '7zip-bin', platformSubdir(process.platform, arch)))
  candidates.push(path7za)
  candidates.push(path.join(app.getAppPath(), 'node_modules', '7zip-bin', platformSubdir(process.platform, arch)))
  for (const c of candidates) {
    if (c && fs.existsSync(c)) return c
  }
  return path7za
}

function platformSubdir(platform: NodeJS.Platform, arch: string): string {
  const os = platform === 'win32' ? 'win' : platform === 'darwin' ? 'mac' : 'linux'
  const a = arch === 'x64' ? 'x64' : 'x86'
  const bin = platform === 'win32' ? '7za.exe' : '7za'
  return path.join(os, a, bin)
}







export async function extractArchive(archivePath: string, extractRoot: string): Promise<string> {
  debugLog('extractArchive:', archivePath, '->', extractRoot)
  if (isZipFormat(archivePath)) {
    extractZipSync(archivePath, extractRoot)
  } else {
    await extractWith7z(archivePath, extractRoot)
  }
  const root = locateGmpRoot(extractRoot)
  debugLog('extractArchive done, gmp root =', root)
  return root
}





export async function extractArchiveAsync(archivePath: string, extractRoot: string): Promise<string> {
  debugLog('extractArchiveAsync:', archivePath, '->', extractRoot)
  if (isZipFormat(archivePath)) {
    fs.mkdirSync(extractRoot, { recursive: true })
    extractZipSync(archivePath, extractRoot)
  } else {
    await new Promise<void>((resolve, reject) => {
      const child = spawn(resolve7za(), ['x', archivePath, `-o${extractRoot}`, '-y'], {
        windowsHide: true,
        stdio: 'ignore'
      })
      child.on('error', (err) => reject(new Error(`7za failed to start: ${err.message}`)))
      child.on('close', (code) => {
        if (code === 0) resolve()
        else reject(new Error(`Extract failed (7za exit ${code}): ${archivePath}`))
      })
    })
  }
  verifyInsideRoot(extractRoot)
  const root = locateGmpRoot(extractRoot)
  debugLog('extractArchiveAsync done, gmp root =', root)
  return root
}

function extractZipSync(archivePath: string, extractRoot: string): void {
  const zip = new AdmZip(archivePath)
  const entries = zip.getEntries()
  for (const entry of entries) {
    const target = safeZipEntryPath(extractRoot, entry.entryName)
    if (entry.isDirectory) {
      fs.mkdirSync(target, { recursive: true })
    } else {
      fs.mkdirSync(path.dirname(target), { recursive: true })
      fs.writeFileSync(target, entry.getData())
    }
  }
}

async function extractWith7z(archivePath: string, extractRoot: string): Promise<void> {
  fs.mkdirSync(extractRoot, { recursive: true })
  await new Promise<void>((resolve, reject) => {
    const child = spawn(resolve7za(), ['x', archivePath, `-o${extractRoot}`, '-y'], {
      windowsHide: true,
      stdio: 'ignore'
    })
    child.on('error', (err) => reject(new Error(`7za failed to start: ${err.message}`)))
    child.on('close', (code) => {
      if (code === 0) resolve()
      else reject(new Error(`Extract failed (7za exit ${code}): ${archivePath}`))
    })
  })
  
  verifyInsideRoot(extractRoot)
}


function verifyInsideRoot(root: string): void {
  const walk = (dir: string): void => {
    let entries: fs.Dirent[]
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true })
    } catch {
      return
    }
    for (const e of entries) {
      const full = path.join(dir, e.name)
      if (!isInside(root, full)) {
        fs.rmSync(root, { recursive: true, force: true })
        throw new Error(`Unsafe archive entry (outside root): ${full}`)
      }
      if (e.isDirectory()) walk(full)
    }
  }
  walk(root)
}





function locateGmpRoot(extractRoot: string): string {
  const candidates: string[] = []
  const walk = (dir: string): void => {
    let found = false
    for (const name of fs.readdirSync(dir)) {
      const p = path.join(dir, name)
      if (name === GMP_METADATA_FOLDER || name === GMP_FALLBACK_METADATA_FOLDER) {
        candidates.push(dir)
        found = true
      }
      if (fs.statSync(p).isDirectory() && !found) {
        walk(p)
      }
    }
  }
  walk(extractRoot)

  
  if (candidates.includes(extractRoot)) return extractRoot
  if (candidates.length === 0) return extractRoot

  
  const sorted = [...candidates].sort((a, b) => a.split(path.sep).length - b.split(path.sep).length)
  const inner = sorted[0]
  if (inner === extractRoot) return extractRoot
  pivotUp(inner, extractRoot)
  return extractRoot
}

function pivotUp(inner: string, extractRoot: string): void {
  
  for (const name of fs.readdirSync(inner)) {
    if (name === GMP_METADATA_FOLDER || name === GMP_FALLBACK_METADATA_FOLDER) continue
    const src = path.join(inner, name)
    const dst = path.join(extractRoot, name)
    fs.renameSync(src, dst)
  }
}


export function createTempDir(gameRoot: string): string {
  const tempBase = path.join(gameRoot, GMP_METADATA_FOLDER, TEMP_FOLDER)
  const dir = path.join(tempBase, `${Date.now()}_${Math.random().toString(36).slice(2, 8)}`)
  fs.mkdirSync(dir, { recursive: true })
  return dir
}

export function removeDirIfInside(root: string, target: string): void {
  if (!isInside(root, target)) {
    throw new Error(`refusing to remove path outside root: ${target}`)
  }
  if (fs.existsSync(target)) {
    fs.rmSync(target, { recursive: true, force: true })
  }
}

File: RBBPMM.Wpf/RBBPMM.Core/ModArchiveExtractor.cs
using System.Diagnostics;
using System.IO.Compression;

namespace RBBPMM.Core;

/// <summary>
/// Archive extraction with path-traversal protection. Mirrors
/// src/main/services/ModArchiveExtractor.ts (zip via System.IO.Compression, everything else via 7z).
/// </summary>
public static class ModArchiveExtractor
{
    /// <summary>Override the 7z executable location (defaults to common install paths + PATH).</summary>
    public static string? SevenZipExecutable { get; set; }

    private static readonly string[] ZipExtensions = [".zip", ".jar", ".apk", ".bbmod", ".gmp"];

    public static bool IsZipFormat(string archivePath)
    {
        var lower = archivePath.ToLowerInvariant();
        if (ZipExtensions.Any(lower.EndsWith))
            return true;
        return StartsWithZipMagic(archivePath);
    }

    public static bool StartsWithZipMagic(string archivePath)
    {
        try
        {
            using var fs = new FileStream(archivePath, FileMode.Open, FileAccess.Read);
            Span<byte> buf = stackalloc byte[4];
            var n = fs.Read(buf);
            return n >= 2 && buf[0] == 0x50 && buf[1] == 0x4B;
        }
        catch
        {
            return false;
        }
    }

    private static string ResolveSevenZip()
    {
        if (!string.IsNullOrEmpty(SevenZipExecutable) && File.Exists(SevenZipExecutable))
            return SevenZipExecutable;
        var candidates = new List<string>
        {
            Path.Combine(AppContext.BaseDirectory, "7z", "7z.exe"),
            Path.Combine(AppContext.BaseDirectory, "resources", "7z", "7z.exe"),
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "Microsoft", "WindowsApps", "7z.exe"),
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), "7-Zip", "7z.exe"),
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86), "7-Zip", "7z.exe"),
        };
        foreach (var c in candidates)
            if (!string.IsNullOrEmpty(c) && File.Exists(c))
                return c;
        return "7z";
    }

    public static async Task<string> ExtractArchiveAsync(string archivePath, string extractRoot)
    {
        if (IsZipFormat(archivePath))
        {
            ExtractZip(archivePath, extractRoot);
        }
        else
        {
            await RunSevenZipAsync(archivePath, extractRoot);
        }
        var root = LocateGmpRoot(extractRoot);
        return root;
    }

    public static string ExtractArchive(string archivePath, string extractRoot)
        => ExtractArchiveAsync(archivePath, extractRoot).GetAwaiter().GetResult();

    public static void ExtractZip(string archivePath, string extractRoot)
    {
        Directory.CreateDirectory(extractRoot);
        using var archive = ZipFile.OpenRead(archivePath);
        foreach (var entry in archive.Entries)
        {
            var target = PathGuard.SafeZipEntryPath(extractRoot, entry.FullName);
            if (entry.FullName.EndsWith('/') || entry.FullName.EndsWith('\\'))
            {
                Directory.CreateDirectory(target);
                continue;
            }
            Directory.CreateDirectory(Path.GetDirectoryName(target)!);
            entry.ExtractToFile(target, overwrite: true);
        }
        VerifyInsideRoot(extractRoot);
    }

    private static async Task RunSevenZipAsync(string archivePath, string extractRoot)
    {
        Directory.CreateDirectory(extractRoot);
        var psi = new ProcessStartInfo
        {
            FileName = ResolveSevenZip(),
            Arguments = $"x \"{archivePath}\" -o\"{extractRoot}\" -y",
            WindowStyle = ProcessWindowStyle.Hidden,
            CreateNoWindow = true,
            UseShellExecute = false,
            RedirectStandardOutput = false,
            RedirectStandardError = false
        };
        using var proc = Process.Start(psi)
            ?? throw new InvalidOperationException("Failed to start 7z");
        await proc.WaitForExitAsync();
        if (proc.ExitCode != 0)
            throw new InvalidOperationException($"Extract failed (7z exit {proc.ExitCode}): {archivePath}");
        VerifyInsideRoot(extractRoot);
    }

    private static void VerifyInsideRoot(string root)
    {
        var rootFull = Path.GetFullPath(root);
        foreach (var file in Directory.EnumerateFileSystemEntries(root, "*", SearchOption.AllDirectories))
        {
            if (!PathGuard.IsInside(rootFull, Path.GetFullPath(file)))
            {
                try { Directory.Delete(rootFull, recursive: true); } catch { /* best effort */ }
                throw new InvalidOperationException($"Unsafe archive entry (outside root): {file}");
            }
        }
    }

    private static string LocateGmpRoot(string extractRoot)
    {
        var rootFull = Path.GetFullPath(extractRoot);
        var candidates = new List<string>();
        CollectGmpCandidateDirs(rootFull, candidates);
        if (candidates.Contains(rootFull))
            return rootFull;
        if (candidates.Count == 0)
            return rootFull;
        var sorted = candidates
            .OrderBy(c => c.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar).Length)
            .ToList();
        var inner = sorted[0];
        if (inner == rootFull)
            return rootFull;
        PivotUp(inner, rootFull);
        return rootFull;
    }

    private static void CollectGmpCandidateDirs(string dir, List<string> outDirs)
    {
        try
        {
            foreach (var name in Directory.GetFileSystemEntries(dir))
            {
                var isDir = Directory.Exists(name);
                var baseName = Path.GetFileName(name);
                if (baseName == Constants.GmpMetadataFolder || baseName == Constants.GmpFallbackMetadataFolder)
                {
                    outDirs.Add(dir);
                    continue;
                }
                if (isDir)
                    CollectGmpCandidateDirs(name, outDirs);
            }
        }
        catch
        {
            // ignore inaccessible entries
        }
    }

    private static void PivotUp(string inner, string extractRoot)
    {
        foreach (var name in Directory.GetFileSystemEntries(inner))
        {
            var baseName = Path.GetFileName(name);
            if (baseName == Constants.GmpMetadataFolder || baseName == Constants.GmpFallbackMetadataFolder)
                continue;
            var dst = Path.Combine(extractRoot, baseName);
            Directory.Move(name, dst);
        }
    }

    public static string CreateTempDir(string gameRoot)
    {
        var tempBase = Path.Combine(gameRoot, Constants.GmpMetadataFolder, Constants.TempFolder);
        var dir = Path.Combine(tempBase, $"{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_{Guid.NewGuid():N}");
        Directory.CreateDirectory(dir);
        return dir;
    }

    public static void RemoveDirIfInside(string root, string target)
    {
        if (!PathGuard.IsInside(Path.GetFullPath(root), Path.GetFullPath(target)))
            throw new InvalidOperationException($"refusing to remove path outside root: {target}");
        if (Directory.Exists(target))
            Directory.Delete(target, recursive: true);
    }
}
